test/distributed/test_nvshmem.py

```diff
@@ -0,0 +1,107 @@
+# Owner(s): ["oncall: distributed"]
+
+# To run:
+# TORCH_SYMMMEM=NVSHMEM python test/distributed/test_nvshmem.py
+# OR
+# TORCH_SYMMMEM=NVSHMEM torchrun --nproc-per-node 4 test/distributed/test_nvshmem.py
+
+import os
+import sys
+import tempfile
+
+import torch
+import torch.distributed as dist
+import torch.distributed._symmetric_memory as symm_mem
+from torch.testing._internal.common_distributed import (
+    MultiProcContinousTest,
+    TEST_SKIPS,
+)
+from torch.testing._internal.common_utils import (
+    skip_but_pass_in_sandcastle_if,
+    skipIfRocm,
+)
+
+
+# Decorator
+# This test requires setting `TORCH_SYMMMEM=NVSHMEM`
+def requires_nvshmem():
+    symm_mem_backend = os.getenv("TORCH_SYMMMEM")
+    return skip_but_pass_in_sandcastle_if(
+        symm_mem_backend != "NVSHMEM",
+        "test_nvshmem requires setting `TORCH_SYMMMEM=NVSHMEM`",
+    )
+
+
+# So that tests are written in device-agnostic way
+device_type = "cuda"
+device_module = torch.get_device_module(device_type)
+
+
+@requires_nvshmem()
+class NVSHMEMSymmetricMemoryTest(MultiProcContinousTest):
+    def setUp(self) -> None:
+        super().setUp()
+        # TODO: relieve this (seems to hang if without)
+        device_module.set_device(self.device)
+        # NOTE: required for nvshmem allocation
+        torch.empty(1, device=self.device)
+
+    # Required by MultiProcContinousTest
+    @classmethod
+    def backend_str(cls) -> str:
+        return "nccl"
+
+    @property
+    def world_size(self) -> int:
+        return device_module.device_count()
+
+    @property
+    def device(self) -> torch.device:
+        return torch.device(device_type, self.rank)
+
+    @skipIfRocm
+    def test_nvshmem_all_to_all(self) -> None:
+        group_name = dist.group.WORLD.group_name
+        symm_mem.enable_symm_mem_for_group(group_name)
```

Comment:
this will be called when you call rendezvous, no?

Reply:
I guess so.